### DISCUSSION TITLE
Canvas faking single pixel

### DISCUSSION THREAD

**Discussion by CookingRoberta** [kkapsner/CanvasBlocker#514] [Ideas]
I am a novice to canvas fingerprinting and to JavaScript as well. Still I wanted to review one possible way of faking a single non isolated pixel of a canvas to a very similar value say (0,0,0) black should become (0,0,1). Similar for transparent pixels.

According to the document "FP-Scanner: The Privacy Implications of BrowserFingerprint Inconsistencies" I found in https://github.com/kkapsner/CanvasBlocker/issues/206 this kind of faking is impossible to detect.

Is this doable, and smart? Or I am missing something, maybe they use single image and fragment that image into 32 smaller images to fingerprint for each part. This way this simple technique cannot work.

**Comment by kkapsner** [comment]
I'm not completely sure what you want to suggest. What do you mean by "non isolated pixel"?

CB already changes the pixels to slightly different colours.

**Reply by CookingRoberta** [reply]
In https://hal.inria.fr/hal-01820197/document "isolated pixels", are pixels whose rgba value is different than(0,0,0,0) and are only surrounded by transparent pixels. In the rendered image, we should not find this kind of pixel because shapes or texts drawn are closed. Non isolated pixel is opposite. These are pixels we can fake.

So this is a proposal to fake single pixel only that is not isolated. I am not 100% sure but after checking the code, but I think CB fakes multiple pixels. 
Since I believe the hash will be different just by faking a single pixels I thought having this single pixel mode would be possible.
Also I am not sure what will happen if someone decides to fake single image fragment by fragment by asking the hash of each fragment (say 32 fragments). This is why it may be smart to fake multiple pixels.

**Reply by kkapsner** [reply]
Yes, CB fakes multiple pixels. Faking only one pixel changes the hash if you hash the complete image.

I also think faking multiple pixels is better.
